src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -392,11 +398,20 @@ fn rustdoc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
     }
 
     let name = unit.pkg.name().to_string();
-    let desc = rustdoc.to_string();
+    let build_state = cx.build_state.clone();
+    let current_id = unit.pkg.package_id().clone();
     let exec_engine = cx.exec_engine.clone();
 
     Ok(Work::new(move |desc_tx| {
-        desc_tx.send(desc).unwrap();
+        let outputs = build_state.outputs.lock().unwrap();
+        for (&(ref id, _), output) in outputs.iter() {
```

Comment:
You'll want to do a hash lookup here to ensure that you're doing the right thing in a cross-compilation scenario. Right now this ignores the `Kind` which is an important distinction for target/host compiles.

The `Kind` to use in the key should be present in `unit.kind` right before the work is created, and that should do the trick!